project_wip_timesheet/wip_journal_entries.py

```diff
@@ -0,0 +1,262 @@
+        """Update the wip journal entry."""
+        if self._is_wip_account_move_reconciled():
+            raise ValidationError(_(
+                'The timesheet line {description} (task: {task}) can not '
+                'be updated because the work in progress entry ({move_name}) is already '
+                'transfered into the cost of goods sold.'
+            ).format(
+                description=self.name,
+                task=str(self.task_id.id),
+                move_name=self.salary_account_move_id.name,
+            ))
+
+        self.salary_account_move_id.state = 'draft'
+        vals = self._get_wip_account_move_vals()
+        self.salary_account_move_id.write(vals)
+        self.salary_account_move_id.post()
+
+    def _reverse_salary_account_move_for_updated_timesheet(self):
+        """Reverse the wip journal entry in the context of an updated timesheet."""
+        if self._is_wip_account_move_reconciled():
+            raise ValidationError(_(
+                'The timesheet line {description} (task: {task}) can not '
+                'be updated because the work in progress entry ({move_name}) would be '
+                'reversed. This journal entry was already transfered into '
+                'the cost of goods sold.'
+            ).format(
+                description=self.name,
+                task=str(self.task_id.id),
+                move_name=self.salary_account_move_id.name,
+            ))
+        self.salary_account_move_id.reverse_moves()
+        self.salary_account_move_id = False
+
+    def _create_update_or_reverse_wip_account_move(self):
+        """Create / Update / Reverse the wip account move.
+
+        Depending on the status of the timesheet line,
+        the wip move is either created, updated or reversed.
+        """
+        must_create_salary_move = (
+            self._requires_wip_salary_move() and not self.salary_account_move_id
+        )
+        must_update_salary_move = (
+            self._requires_wip_salary_move() and self.salary_account_move_id
+        )
+        must_reverse_salary_move = (
+            not self._requires_wip_salary_move() and self.salary_account_move_id
+        )
+
+        if must_create_salary_move:
+            self._create_wip_account_move()
+
+        elif must_update_salary_move:
+            self._update_wip_account_move()
+
+        elif must_reverse_salary_move:
+            self._reverse_salary_account_move_for_updated_timesheet()
+
+    @api.model
+    def create(self, vals):
+        """On timesheet create, create or update the wip journal entry.
+
+        Because the creation of a timesheet is complex, the journal entry
+        may be created by a write before the return of super().create(vals).
+        """
+        line = super().create(vals)
+        if line._requires_wip_salary_move():
+            line.sudo()._create_update_or_reverse_wip_account_move()
+        return line
+
+    def _get_salary_move_dependent_fields(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3307196)